### PR TITLE
add cache scale and update

### PR DESCRIPTION
**Pull request by Ghostbaby**

### CODE REVIEW

File: controllers/cache/redis.go [hunk 1]
```diff
@@ -65,17 +70,18 @@ func (redis *RedisReconciler) Delete() (*lcm.CRStatus, error) {
...
-	panic("implement me")
-}
-
-func (redis *RedisReconciler) ScaleUp(newReplicas uint64) (*lcm.CRStatus, error) {
-	panic("implement me")
-}
-
-func (redis *RedisReconciler) ScaleDown(newReplicas uint64) (*lcm.CRStatus, error) {
-	panic("implement me")
+	if err := redis.ScaleUpCache(); err != nil {
```
Comment (steven-zou) [inline controllers/cache/redis.go:73]: Don't we need to check function to determine if it's up or down?

File: controllers/cache/redis.go [hunk 2]
```diff
@@ -65,17 +70,18 @@ func (redis *RedisReconciler) Delete() (*lcm.CRStatus, error) {
...
+	}
+	return redis.CRStatus, nil
 }
 
 func (redis *RedisReconciler) Update(spec *goharborv1.HarborCluster) (*lcm.CRStatus, error) {
-	panic("implement me")
+	if err := redis.RollingUpgrades(); err != nil {
+		return redis.CRStatus, err
+	}
+	return redis.CRStatus, nil
```
Comment (steven-zou) [inline controllers/cache/redis.go:86]: move `CRStatus`  out of reconciler like what we did for postgresql?

File: controllers/cache/scale_up.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+	corev1 "k8s.io/api/core/v1"
+)
+
+// UpScale reconcile will upscale Redis sentinel cluster if replicas upscale.
+// It does:
+// - check replicas
+// - update RedisFailovers CR replicas
+func (redis *RedisReconciler) ScaleUpCache() error {
+
+	crdClient := redis.DClient.WithResource(redisFailoversGVR).WithNamespace(redis.Namespace)
```
Comment (steven-zou) [inline controllers/cache/scale_up.go:20]: It seems there are some similar codes in the `ScaleUpCache` and `ScaleDownCache` methods, is it possible to reorganize code to reduce duplicated parts?

File: controllers/cache/constants.go [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package cache
+
+const (
+	GetRedisClientError = "fail to get redis client"
```
Comment (soulseen) [inline controllers/cache/constants.go:4]: format😄??

File: controllers/cache/provision.go [hunk 5]
```diff
@@ -31,38 +32,38 @@ func (redis *RedisReconciler) Deploy() error {
...
 		}
 
 		redis.Log.Info("Creating Redis.", "namespace", redis.Namespace, "name", redis.Name)
 		_, err = crdClient.Create(expectCR, metav1.CreateOptions{})
 		if err != nil {
-			return err
+			return cacheNotReadyStatus(CreateRedisCrError, err.Error()),err
 		}
 
 		redis.Log.Info("Redis create complete.", "namespace", redis.Namespace, "name", redis.Name)
```
Comment (soulseen) [inline controllers/cache/provision.go:55]: here just create Redis，not completed.

File: controllers/cache/provision.go [hunk 6]
```diff
@@ -18,10 +19,10 @@ import (
 // - perform any RedisFailovers downscale (left for downscale phase)
 // - perform any RedisFailovers upscale (left for upscale phase)
 // - perform any pod upgrade (left for rolling upgrade phase)
-func (redis *RedisReconciler) Deploy() error {
+func (redis *RedisReconciler) Deploy() (*lcm.CRStatus, error) {
 
 	if redis.HarborCluster.Spec.Redis.Kind == "external" {
-		return nil
+		return cacheUnknownStatus(),nil
```
Comment (soulseen) [inline controllers/cache/provision.go:25]: why return `UnknownStatus` when using `external mode`?

**Comment by soulseen** [comment]
please resolve conflicts.

File: controllers/cache/redis.go (hunk 1 above)
Comment (Ghostbaby) [reply controllers/cache/redis.go:74]: We need to do different operations on different actions. If scale down, we need to check whether the node to be destroyed is the master node.

File: controllers/cache/provision.go (hunk 6 above)
Comment (Ghostbaby) [reply controllers/cache/provision.go:25]: Reconcile will flow to the `Readiness` controller, `Readiness` will use ping to  check external Redis.

File: controllers/cache/constants.go [hunk 7]
```diff
@@ -0,0 +1,20 @@
+package cache
+
+const (
+	GetRedisClientError = "Get redis client error"
+	CheckRedisHealthError = "Check redis health error"
+	CreateComponentSecretError = "Create component secret error"
+
+	GenerateRedisCrError = "Generate redis cr error"
```
Comment (steven-zou) [inline controllers/cache/constants.go:8]:
```suggestion
	GenerateRedisCrError = "Generate redis CR error"
```

I think all `cr` should be changed to `CR`

File: controllers/cache/redis.go [hunk 8]
```diff
@@ -43,39 +39,52 @@ func (redis *RedisReconciler) Reconcile() (*lcm.CRStatus, error) {
...
-	panic("implement me")
-}
-
-func (redis *RedisReconciler) ScaleUp(newReplicas uint64) (*lcm.CRStatus, error) {
-	panic("implement me")
-}
-
-func (redis *RedisReconciler) ScaleDown(newReplicas uint64) (*lcm.CRStatus, error) {
-	panic("implement me")
+	crStatus, err := redis.ScaleUpCache()
```
Comment (steven-zou) [inline controllers/cache/redis.go:73]: I think we should not use such "testing" way to do scale up or down/

We should compare the spec to know what kind of scaling we should do first, and then do the right thing.

File: controllers/cache/scale_down.go [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+	downscales := calculateDownscales(expectCR, actualCR)
+	leavingNodes := leavingNodeNames(downscales)
+
+	if len(leavingNodes) != 0 {
+		redis.Log.Info("Migrating data away from nodes", "nodes", leavingNodes)
+	}
+
+	if expectReplica < actualReplica {
+
+		_, redisPodList, err := redis.GetStatefulSetPods()
```
Comment (steven-zou) [inline controllers/cache/scale_down.go:60]: My question why we need to do such lower operations to do scaling of redis? Does the Redis CR not supporting updating process? e.g.: we just change the replicas of the Redis server/sentinel servers and the redis CR reconcile will help to do the work.

File: controllers/cache/constants.go (hunk 7 above)
Comment (steven-zou) [reply controllers/cache/constants.go:8]: @Ghostbaby 

cr->CR?

File: controllers/cache/labels.go [hunk 10]
```diff
@@ -7,10 +7,10 @@ const (
 // NewLabels returns new labels
 func (redis *RedisReconciler) NewLabels() map[string]string {
 	dynLabels := map[string]string{
-		AppLabel:                     redis.Name,
+		AppLabel:                     redis.HarborCluster.Name,
```
Comment (steven-zou) [inline controllers/cache/labels.go:10]: AppLabel has similar value with `"goharbor.io/harbor-cluster"`?

File: controllers/cache/provision.go [hunk 11]
```diff
@@ -18,51 +19,38 @@ import (
...
-		return err
+		return cacheNotReadyStatus(CreateRedisCrError, err.Error()), err
 	}
 
-	redis.ExpectCR = expectCR
-	redis.ActualCR = actualCR
-
-	return nil
+	redis.Log.Info("Redis has been created.", "namespace", redis.HarborCluster.Namespace, "name", redis.HarborCluster.Name)
+	return cacheUnknownStatus(), nil
```
Comment (steven-zou) [inline controllers/cache/provision.go:53]: When the status can be changed to be the "ready" status from this `cacheUnknownStatus`? After `Readiness()` confirmation?

Comment (Ghostbaby) [reply controllers/cache/provision.go:53]: Yes, `Readiness()` will check whether Redis is available, and create component secret after passing the check, and finally return to the ready state.

**Pull request merged**